cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -164,6 +169,50 @@ public TwillController launch(TwillApplication twillApplication) {
   }
 
   /**
+   * Gets plugin files that needs to be localized for the adapter. If the given run is not an adapter, no
+   * modification will be done to the map.
+   */
+  private Map<String, File> addAdapterPluginFiles(ProgramOptions options, Map<String, File> localizeFiles) {
+    Arguments arguments = options.getArguments();
+    if (!arguments.hasOption(ProgramOptionConstants.ADAPTER_SPEC)) {
+      return localizeFiles;
+    }
+
+    // Decode the adapter spec from program system argument
+    AdapterSpecification adapterSpec = GSON.fromJson(arguments.getOption(ProgramOptionConstants.ADAPTER_SPEC),
+                                                     AdapterSpecification.class);
+
+    // Get all unique PluginInfo from the adapter spec
+    Set<PluginInfo> plugins = adapterSpec.getPluginInfos();
+
+    // If there is no plugin used by the adapter, nothing need to be localized
+    if (plugins.isEmpty()) {
+      return localizeFiles;
+    }
+
+    File templateDir = new File(cConf.get(Constants.AppFabric.APP_TEMPLATE_DIR));
```

Comment:
The `PluginInstantiator` works with this dir structure, hence maintaining the dir structure make the PluginInstantiator much simpler to be used in different mode.